wasm/derive/src/validator.rs

```diff
@@ -0,0 +1,63 @@
+//! Macro for writing validator entrypoint
+
+#![allow(clippy::panic)]
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, parse_quote};
+
+/// [`validator_entrypoint`](crate::validator_entrypoint()) macro implementation
+#[allow(clippy::needless_pass_by_value)]
+pub fn impl_entrypoint(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let syn::ItemFn {
+        attrs,
+        vis,
+        sig,
+        mut block,
+    } = parse_macro_input!(item);
+
+    let fn_name = &sig.ident;
+    assert!(
+        matches!(sig.output, syn::ReturnType::Type(_, _)),
+        "Validator entrypoint must have `Verdict` return type"
+    );
+
+    let arg: syn::Expr = parse_quote! {{
+        let top_operation = ::iroha_wasm::query_operation_to_validate();
+        ::core::convert::TryInto::try_into(top_operation)
+            .dbg_expect("Failed to convert top-level operation to the concrete one")
+    }};
+
+    block.stmts.insert(
+        0,
+        parse_quote!(
+            use ::iroha_wasm::Execute as _;
+        ),
+    );
+
+    quote! {
+        /// Validator entrypoint
+        ///
+        /// # Memory safety
+        ///
+        /// This function transfers the ownership of allocated
+        /// [`Verdict`](::iroha_wasm::data_model::permission::validator::Verdict)
+        #[no_mangle]
+        pub unsafe extern "C" fn _iroha_validator_main()
+            -> *const u8
+        {
+            use ::iroha_wasm::DebugExpectExt as _;
+
+            let verdict: ::iroha_wasm::data_model::permission::validator::Verdict = #fn_name(#arg);
+            let bytes_box =
```

Comment:
I'm not really sure that we could pass ownership this way, i suspect that smartcontract memory is not the same memory used by host iroha, this probaly explains error you recieve, when you try to wrap pointer into `Box` on the other end, you can't deallocate 
memory that was allocated by other allocator.
IMO you actually could use `Box::leak` here to recieve pointer, you would anyway free memory after droping runtime, however this should be investigated.

Reply:
I understand your suspection. But we are already using one allocator to allocate memory and another allocator to deallocate. It's used when Iroha allocates memory for i.e. query result and when WASM deallocates it. But somehow it's not woking on the contrary.
I've provided a possible solution in the PR description: allocate and deallocate memory for `Verdict` only on Iroha side

Reply:
`Box::leak()` leads to the same thing as `ManuallyDrop` here. I just use the approach that was used in another places, see `wasm/src/lib.rs`

Reply:
> It's used when Iroha allocates memory for i.e. query result and when WASM deallocates it.

@Arjentix could you please point me where this happens?

Reply:
Oh, I was wrong. On Iroha side we allocate with exported `_iroha_wasm_alloc()` (look at `encode_bytes_into_memory()`)

Reply:
Looks like I need to create `_iroha_wasm_dealloc()` export and use it

Reply:
Done